File: iot-backend/import_dataset.py
#!/usr/bin/env python3
# ============================================================
#  ABBKA - Import Dataset dans MongoDB
# ============================================================

import pandas as pd
from pymongo import MongoClient
from datetime import datetime
import numpy as np

# ==================== CONFIGURATION ====================
MONGO_URI = "mongodb://localhost:27017/"
DATABASE_NAME = "abbka"
COLLECTION_NAME = "training_data"

# ==================== CONNEXION MONGODB ====================
client = MongoClient(MONGO_URI)
db = client[DATABASE_NAME]
collection = db[COLLECTION_NAME]

print("╔════════════════════════════════════════════════════════╗")
print("║       ABBKA - Import Dataset Training                 ║")
print("╚════════════════════════════════════════════════════════╝\n")

# ==================== CHARGER DATASET ====================

print("[1/4] Chargement du fichier CSV...")
df = pd.read_csv('ai4i2020.csv')
print(f"  ✅ {len(df)} lignes chargées\n")

print("[2/4] Aperçu des données :")
print(df.head())
print(f"\nColonnes : {list(df.columns)}\n")

# ==================== NETTOYER DONNÉES ====================
print("[3/4] Nettoyage des données...")
collection.delete_many({})
print(f"  🗑️  Collection vidée")

# ==================== TRANSFORMATION ====================
print("\n[4/4] Insertion dans MongoDB...")

records = []
for idx, row in df.iterrows():
    record = {
        "udi": int(row.get('UDI', idx)),
        "product_id": row.get('Product ID', f'PROD_{idx}'),
        "type": row.get('Type', 'M'),
        "metrics": {
            "thermal": round(float(row.get('Air temperature [K]', 300)) - 273.15, 1),
            "process_temp": round(float(row.get('Process temperature [K]', 310)) - 273.15, 1),
            "rotational_speed": int(row.get('Rotational speed [rpm]', 1500)),
            "torque": round(float(row.get('Torque [Nm]', 40)), 2),
            "tool_wear": int(row.get('Tool wear [min]', 0)),
            "pressure": round(float(row.get('Torque [Nm]', 40)) / 10.0, 2),
            "power": round(float(row.get('Rotational speed [rpm]', 1500)) / 20.0, 1),
        },
        "failure": {
            "machine_failure": int(row.get('Machine failure', 0)),
            "twf": int(row.get('TWF', 0)),
            "hdf": int(row.get('HDF', 0)),
            "pwf": int(row.get('PWF', 0)),
            "osf": int(row.get('OSF', 0)),
            "rnf": int(row.get('RNF', 0))
        },
        "timestamp": datetime.utcnow(),
        "source": "AI4I_2020_Dataset",
        "imported_at": datetime.utcnow()
    }
    records.append(record)

    if len(records) >= 1000:
        collection.insert_many(records)
        print(f"  📥 {idx+1} enregistrements insérés...")
        records = []

if records:
    collection.insert_many(records)
    print(f"  📥 {len(records)} enregistrements finaux insérés...")

# ==================== VÉRIFICATION ====================
total_count = collection.count_documents({})
failure_count = collection.count_documents({"failure.machine_failure": 1})

print(f"\n╔════════════════════════════════════════════════════════╗")
print(f"║  ✅ IMPORT TERMINÉ                                     ║")
print(f"║  📊 Total enregistrements : {total_count:<8}                  ║")
print(f"║  ⚠️  Pannes détectées     : {failure_count:<8} ({failure_count/total_count*100:.1f}%)           ║")
print(f"╚════════════════════════════════════════════════════════╝")

# ==================== CRÉER INDEXES ====================
print("\n[INDEX] Création des index MongoDB...")
collection.create_index("udi")
collection.create_index("timestamp")
collection.create_index("failure.machine_failure")
print("  ✅ Index créés\n")

# ==================== STATISTIQUES ====================
print("[STATS] Statistiques du dataset :")
pipeline = [
    {
        "$group": {
            "_id": "$type",
            "count": {"$sum": 1},
            "avg_temp": {"$avg": "$metrics.thermal"},
            "failure_rate": {"$avg": "$failure.machine_failure"}
        }
    }
]
stats = list(collection.aggregate(pipeline))
for stat in stats:
    print(f"  Type {stat['_id']} : {stat['count']} samples, "
          f"Temp moy: {stat['avg_temp']:.1f}°C, "
          f"Taux panne: {stat['failure_rate']*100:.1f}%")

print("\n✅ Prêt pour l'entraînement IA !\n")

File: iot-backend/import_dataset.js
// ============================================================
//  ABBKA - Import Dataset AI4I 2020 dans MongoDB (Node.js)
//  Fichier : import_dataset.js
// ============================================================

const fs = require('fs');
const path = require('path');
const mongoose = require('mongoose');
const readline = require('readline');

const MONGO_URI = 'mongodb://localhost:27017/abbka';
const COLLECTION_NAME = 'training_data';
const CSV_FILE = path.join(__dirname, 'ai4i2020.csv');

// ==================== SCHEMA ====================
const trainingSchema = new mongoose.Schema({
    udi: Number,
    product_id: String,
    type: String,
    metrics: {
        thermal: Number,
        process_temp: Number,
        rotational_speed: Number,
        torque: Number,
        tool_wear: Number,
        pressure: Number,
        power: Number
    },
    failure: {
        machine_failure: Number,
        twf: Number,
        hdf: Number,
        pwf: Number,
        osf: Number,
        rnf: Number
    },
    timestamp: Date,
    source: String,
    imported_at: Date
});

const Training = mongoose.model('training_data', trainingSchema, COLLECTION_NAME);

// ==================== MAIN ====================
async function main() {
    console.log('╔════════════════════════════════════════════════════════╗');
    console.log('║       ABBKA - Import Dataset Training (Node.js)       ║');
    console.log('╚════════════════════════════════════════════════════════╝\n');

    if (!fs.existsSync(CSV_FILE)) {
        console.error('❌ Fichier ai4i2020.csv introuvable dans:', __dirname);
        console.error('\n📥 Télécharger le dataset :');
        console.error('   https://archive.ics.uci.edu/ml/machine-learning-databases/00601/ai4i2020.csv');
        console.error('   Ou Kaggle: https://www.kaggle.com/datasets/stephanmatzka/predictive-maintenance-dataset-ai4i-2020\n');
        console.error('   Placez le fichier dans:', __dirname);
        process.exit(1);
    }

    // Connexion MongoDB
    await mongoose.connect(MONGO_URI);
    console.log('📦 MongoDB connecté');

    // Vider la collection
    await Training.deleteMany({});
    console.log('🗑️  Collection vidée\n');

    // Lire le CSV — collecter toutes les lignes d'abord
    const lines = await new Promise((resolve, reject) => {
        const result = [];
        const rl = readline.createInterface({ input: fs.createReadStream(CSV_FILE), crlfDelay: Infinity });
        rl.on('line', line => { if (line.trim()) result.push(line); });
        rl.on('close', () => resolve(result));
        rl.on('error', reject);
    });

    let headers = null;
    let records = [];
    let lineCount = 0;
    let totalInserted = 0;

    for (const line of lines) {
        if (!line.trim()) continue;
        const cols = line.split(',').map(c => c.trim().replace(/^"|"$/g, ''));

        if (!headers) {
            headers = cols;
            console.log('[1/4] En-têtes CSV :', headers.join(', '));
            continue;
        }

        const row = {};
        headers.forEach((h, i) => row[h] = cols[i]);

        const record = {
            udi: parseInt(row['UDI'] || lineCount),
            product_id: row['Product ID'] || `PROD_${lineCount}`,
            type: row['Type'] || 'M',
            metrics: {
                thermal: parseFloat((parseFloat(row['Air temperature [K]'] || 300) - 273.15).toFixed(1)),
                process_temp: parseFloat((parseFloat(row['Process temperature [K]'] || 310) - 273.15).toFixed(1)),
                rotational_speed: parseInt(row['Rotational speed [rpm]'] || 1500),
                torque: parseFloat(parseFloat(row['Torque [Nm]'] || 40).toFixed(2)),
                tool_wear: parseInt(row['Tool wear [min]'] || 0),
                pressure: parseFloat((parseFloat(row['Torque [Nm]'] || 40) / 10.0).toFixed(2)),
                power: parseFloat((parseFloat(row['Rotational speed [rpm]'] || 1500) / 20.0).toFixed(1))
            },
            failure: {
                machine_failure: parseInt(row['Machine failure'] || 0),
                twf: parseInt(row['TWF'] || 0),
                hdf: parseInt(row['HDF'] || 0),
                pwf: parseInt(row['PWF'] || 0),
                osf: parseInt(row['OSF'] || 0),
                rnf: parseInt(row['RNF'] || 0)
            },
            timestamp: new Date(),
            source: 'AI4I_2020_Dataset',
            imported_at: new Date()
        };

        records.push(record);
        lineCount++;

        if (records.length >= 1000) {
            await Training.insertMany(records);
            totalInserted += records.length;
            process.stdout.write(`\r  📥 ${totalInserted} enregistrements insérés...`);
            records = [];
        }
    }

    if (records.length > 0) {
        await Training.insertMany(records);
        totalInserted += records.length;
    }

    console.log(`\n  📥 Total: ${totalInserted} enregistrements insérés`);

    // ==================== STATS ====================
    const total = await Training.countDocuments();
    const failures = await Training.countDocuments({ 'failure.machine_failure': 1 });
    const pct = ((failures / total) * 100).toFixed(1);

    console.log('\n╔════════════════════════════════════════════════════════╗');
    console.log(`║  ✅ IMPORT TERMINÉ                                     ║`);
    console.log(`║  📊 Total enregistrements : ${String(total).padEnd(8)}                 ║`);
    console.log(`║  ⚠️  Pannes détectées     : ${String(failures).padEnd(8)} (${pct}%)          ║`);
    console.log('╚════════════════════════════════════════════════════════╝');

    // Créer les index
    console.log('\n[INDEX] Création des index MongoDB...');
    await mongoose.connection.collection(COLLECTION_NAME).createIndex({ udi: 1 });
    await mongoose.connection.collection(COLLECTION_NAME).createIndex({ timestamp: 1 });
    await mongoose.connection.collection(COLLECTION_NAME).createIndex({ 'failure.machine_failure': 1 });
    console.log('  ✅ Index créés');

    // Statistiques par type
    console.log('\n[STATS] Répartition par type de machine :');
    const stats = await Training.aggregate([
        {
            $group: {
                _id: '$type',
                count: { $sum: 1 },
                avg_temp: { $avg: '$metrics.thermal' },
                failure_rate: { $avg: '$failure.machine_failure' }
            }
        },
        { $sort: { _id: 1 } }
    ]);

    stats.forEach(s => {
        console.log(`  Type ${s._id} : ${s.count} samples | Temp moy: ${s.avg_temp.toFixed(1)}°C | Taux panne: ${(s.failure_rate * 100).toFixed(1)}%`);
    });

    console.log('\n✅ Prêt pour l\'entraînement IA !\n');
    await mongoose.disconnect();
    process.exit(0);
}

main().catch(err => {
    console.error('❌ Erreur:', err.message);
    process.exit(1);
});
